src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/TagletWriter.java

```diff
@@ -25,226 +25,101 @@
+         * page for a declaration.
+         */
+        public final boolean inSummary;
+        /**
+         * The set of enclosing kinds of tags.
+         */
+        public final Set<DocTree.Kind> inTags;
+
+        /**
+         * Creates an outermost context, with no enclosing tags.
+         *
+         * @param isFirstSentence {@code true} if the trees are appearing in a context of just the
+         *                        first sentence and {@code false} otherwise
+         * @param inSummary       {@code true} if the trees are appearing in the "summary" section
+         *                        of the page for a declaration and {@code false} otherwise
+         */
+        public Context(boolean isFirstSentence, boolean inSummary) {
+            this(isFirstSentence, inSummary, EnumSet.noneOf(DocTree.Kind.class));
+        }
 
-    /**
-     * Returns the output for a {@code {@snippet ...}} tag.
-     *
-     * @param element    The element that owns the doc comment
-     * @param snippetTag the snippet tag
-     * @param id         the value of the id attribute, or null if not defined
-     * @param lang       the value of the lang attribute, or null if not defined
-     *
-     * @return the output
-     */
-    protected abstract Content snippetTagOutput(Element element, SnippetTree snippetTag, StyledText text,
-                                                String id, String lang);
+        private Context(boolean isFirstSentence, boolean inSummary, Set<DocTree.Kind> inTags) {
+            this.isFirstSentence = isFirstSentence;
+            this.inSummary = inSummary;
+            this.inTags = inTags;
+        }
 
-    /**
-     * Returns the output for one or more {@code @spec} tags.
-     *
-     * @param element  the element that owns the doc comment
-     * @param specTags the array of @spec tags.
-     *
-     * @return the output
-     */
-    protected abstract Content specTagOutput(Element element, List<? extends SpecTree> specTags);
+        /**
+         * Creates a new {@code Context} that includes an extra tag kind in the set of enclosing
+         * kinds of tags.
+         *
+         * @param tree the enclosing tree
+         *
+         * @return the new {@code Context}
+         */
+        public Context within(DocTree tree) {
+            var newInTags = EnumSet.copyOf(inTags);
+            newInTags.add(tree.getKind());
+            return new Context(isFirstSentence, inSummary, newInTags);
+        }
+    }
 
     /**
-     * Returns the output for a {@code {@systemProperty...}} tag.
-     *
-     * @param element           the element that owns the doc comment
-     * @param systemPropertyTag the system property tag
-     *
-     * @return the output
+     * The context in which to generate the output for a series of {@code DocTree} nodes.
      */
-    protected abstract Content systemPropertyTagOutput(Element element, SystemPropertyTree systemPropertyTag);
+    public final Context context;
```

Comment:
I dislike it when a variable is accessible both directly and through the getter. Nonencapsulated fields like that is harder to debug and maintain. That said, I realise that it's somewhat of a pattern in jdk.javadoc. So, no need to change it at this time; I just make an observation.

Reply:
Observation noted. For my part, I don't mind non-encapsulated fields if they are final.  I prefer that style to the parens of getter methods:
```
var x = utils.something;
var x = utils().something;
```

Reply:
My comment was about fields that are BOTH accessible directly and through the getter. To me, it's an either-or situation.

Regardless, don't bother with changing it now. I just felt like it's a good opportunity to raise that general concern.